From patchwork Tue Apr 14 00:10:25 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 11486479
Return-Path: <SRS0=cOhZ=56=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id AC9D91392
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 14 Apr 2020 00:10:34 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 8B9522072D
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 14 Apr 2020 00:10:34 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=google.com header.i=@google.com
 header.b="Dxqwa6hO"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728588AbgDNAKc (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 13 Apr 2020 20:10:32 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:44784 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-FAIL-OK-FAIL)
        by vger.kernel.org with ESMTP id S1727878AbgDNAKb (ORCPT
        <rfc822;kvm@vger.kernel.org>); Mon, 13 Apr 2020 20:10:31 -0400
Received: from mail-pf1-x449.google.com (mail-pf1-x449.google.com
 [IPv6:2607:f8b0:4864:20::449])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 0ACE7C0A3BDC
        for <kvm@vger.kernel.org>; Mon, 13 Apr 2020 17:10:31 -0700 (PDT)
Received: by mail-pf1-x449.google.com with SMTP id i128so1908086pfc.4
        for <kvm@vger.kernel.org>; Mon, 13 Apr 2020 17:10:31 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=YRE0bl+MbOZJemQI6/3rM1uBjjYWT+ZjbHxoUQ9wtoQ=;
        b=Dxqwa6hO3h2eMm0jA1G2oQJUpnCnCahrkI6Ohz7isGPExDLuwqXl4Qbt2nov2HzIio
         Pgh+u05iOAzol3TPtEsyKN59KeO6weSaIT0RQmiepTS79N6Qxp/CmXJU+1/mDxtQoMNZ
         8+hvBluchnkRkiBH3tdpoMrGHOg45+QInV2QXElQDLpiq860HePTu/kIihECKGpeNSLq
         PfWtAoavfZR4u9H5rRw3U7lVsPoKUHe83788sF8OfrwSJ2RkItNQgj5jOawBDk2j3WYN
         TNPzk59tEJ8QVY/MLePa/dheKw8Nd7vrBI1kLtrwcAymc8d7gGAppLjFbO/iEFLECcEq
         0gPA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=YRE0bl+MbOZJemQI6/3rM1uBjjYWT+ZjbHxoUQ9wtoQ=;
        b=bldqcfAby/TwcFs+kldBc5mGYgg88Hxxbiu/AVoVG2/+kw1yKPp5tqdsj6ftac8+kx
         kFasignQUptzdhJYllbL7ulo2v9Wq1DIL4JtA6v/2UaGctR0kpBWbY3TsRtI2oxuWFVa
         wmt6B0QBXMn6PQJ62kT3QEzYy8iWe5DKgLZ1jPK9SXTMAoiZswtDeBdDezQZPHT8S+xr
         kr+UDHW/8uEqooBKU4PdoJSd2x9/q1o2uCFrppjn9+0ze45kM/btxlou3U6Ic0HE+bUk
         P3vriV5354yRxV89y74HYZKkDHopq513qa9kH48MgZhB8sWirCIO3SObGZ0dQT/vrCGq
         ZTOw==
X-Gm-Message-State: AGi0PubmSBUmN9wQ5FbAZg7nVwRy4nvOG6L0XheHLsMxgNKHNovn9rQO
        6M6Whp2wdg+Fl390u1GVrc5eLelq+jqi2VmXX6rWNIgNcuH38Qwqh6ocZmHQ2caicemd89gy7XE
        2Zi0rtyONTdJzE+aNDFAy6q/CkhLAFjpze2vvYcpomvg8gmRlapLKj8vTxtwR66w=
X-Google-Smtp-Source: 
 APiQypIDqlhIhVNgUbLZcXUU/8Q+TWQqbZIrN4XOzNxTArCp/mwyWyTlRLEQ1TYSQ2+eeTztkiWuDxFDI/mHsQ==
X-Received: by 2002:a17:90a:46cf:: with SMTP id
 x15mr24370313pjg.77.1586823030511;
 Mon, 13 Apr 2020 17:10:30 -0700 (PDT)
Date: Mon, 13 Apr 2020 17:10:25 -0700
Message-Id: <20200414001026.50051-1-jmattson@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.26.0.110.g2183baf09c-goog
Subject: [kvm-unit-tests PATCH 1/2] x86: nVMX: Add some corner-case
 VMX-preemption timer tests
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>,
        Oliver Upton <oupton@google.com>,
        Peter Shier <pshier@google.com>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Verify that both injected events and debug traps that result from
pending debug exceptions take precedence over a "VMX-preemption timer
expired" VM-exit resulting from a zero-valued VMX-preemption timer.

Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Oliver Upton <oupton@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
---
 x86/vmx_tests.c | 120 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 120 insertions(+)

diff --git a/x86/vmx_tests.c b/x86/vmx_tests.c
index 1f97fe3..fccb27f 100644
--- a/x86/vmx_tests.c
+++ b/x86/vmx_tests.c
@@ -8319,6 +8319,125 @@ static void vmx_store_tsc_test(void)
 	       msr_entry.value, low, high);
 }
 
+static void vmx_preemption_timer_zero_test_db_handler(struct ex_regs *regs)
+{
+}
+
+static void vmx_preemption_timer_zero_test_guest(void)
+{
+	while (vmx_get_test_stage() < 3)
+		vmcall();
+}
+
+static void vmx_preemption_timer_zero_activate_preemption_timer(void)
+{
+	vmcs_set_bits(PIN_CONTROLS, PIN_PREEMPT);
+	vmcs_write(PREEMPT_TIMER_VALUE, 0);
+}
+
+static void vmx_preemption_timer_zero_advance_past_vmcall(void)
+{
+	vmcs_clear_bits(PIN_CONTROLS, PIN_PREEMPT);
+	enter_guest();
+	skip_exit_vmcall();
+}
+
+static void vmx_preemption_timer_zero_inject_db(bool intercept_db)
+{
+	vmx_preemption_timer_zero_activate_preemption_timer();
+	vmcs_write(ENT_INTR_INFO, INTR_INFO_VALID_MASK |
+		   INTR_TYPE_HARD_EXCEPTION | DB_VECTOR);
+	vmcs_write(EXC_BITMAP, intercept_db ? 1 << DB_VECTOR : 0);
+	enter_guest();
+}
+
+static void vmx_preemption_timer_zero_set_pending_dbg(u32 exception_bitmap)
+{
+	vmx_preemption_timer_zero_activate_preemption_timer();
+	vmcs_write(GUEST_PENDING_DEBUG, BIT(12) | DR_TRAP1);
+	vmcs_write(EXC_BITMAP, exception_bitmap);
+	enter_guest();
+}
+
+static void vmx_preemption_timer_zero_expect_preempt_at_rip(u64 expected_rip)
+{
+	u32 reason = (u32)vmcs_read(EXI_REASON);
+	u64 guest_rip = vmcs_read(GUEST_RIP);
+
+	report(reason == VMX_PREEMPT && guest_rip == expected_rip,
+	       "Exit reason is 0x%x (expected 0x%x) and guest RIP is %lx (0x%lx expected).",
+	       reason, VMX_PREEMPT, guest_rip, expected_rip);
+}
+
+/*
+ * This test ensures that when the VMX preemption timer is zero at
+ * VM-entry, a VM-exit occurs after any event injection and after any
+ * pending debug exceptions are raised, but before execution of any
+ * guest instructions.
+ */
+static void vmx_preemption_timer_zero_test(void)
+{
+	u64 db_fault_address = (u64)get_idt_addr(&boot_idt[DB_VECTOR]);
+	handler old_db;
+	u32 reason;
+
+	if (!(ctrl_pin_rev.clr & PIN_PREEMPT)) {
+		report_skip("'Activate VMX-preemption timer' not supported");
+		return;
+	}
+
+	/*
+	 * Install a custom #DB handler that doesn't abort.
+	 */
+	old_db = handle_exception(DB_VECTOR,
+				  vmx_preemption_timer_zero_test_db_handler);
+
+	test_set_guest(vmx_preemption_timer_zero_test_guest);
+
+	/*
+	 * VMX-preemption timer should fire after event injection.
+	 */
+	vmx_set_test_stage(0);
+	vmx_preemption_timer_zero_inject_db(0);
+	vmx_preemption_timer_zero_expect_preempt_at_rip(db_fault_address);
+	vmx_preemption_timer_zero_advance_past_vmcall();
+
+	/*
+	 * VMX-preemption timer should fire after event injection.
+	 * Exception bitmap is irrelevant, since you can't intercept
+	 * an event that you injected.
+	 */
+	vmx_set_test_stage(1);
+	vmx_preemption_timer_zero_inject_db(1 << DB_VECTOR);
+	vmx_preemption_timer_zero_expect_preempt_at_rip(db_fault_address);
+	vmx_preemption_timer_zero_advance_past_vmcall();
+
+	/*
+	 * VMX-preemption timer should fire after pending debug exceptions
+	 * have delivered a #DB trap.
+	 */
+	vmx_set_test_stage(2);
+	vmx_preemption_timer_zero_set_pending_dbg(0);
+	vmx_preemption_timer_zero_expect_preempt_at_rip(db_fault_address);
+	vmx_preemption_timer_zero_advance_past_vmcall();
+
+	/*
+	 * VMX-preemption timer would fire after pending debug exceptions
+	 * have delivered a #DB trap, but in this case, the #DB trap is
+	 * intercepted.
+	 */
+	vmx_set_test_stage(3);
+	vmx_preemption_timer_zero_set_pending_dbg(1 << DB_VECTOR);
+	reason = (u32)vmcs_read(EXI_REASON);
+	report(reason == VMX_EXC_NMI, "Exit reason is 0x%x (expected 0x%x)",
+	       reason, VMX_EXC_NMI);
+
+	vmcs_clear_bits(PIN_CONTROLS, PIN_PREEMPT);
+	enter_guest();
+
+	handle_exception(DB_VECTOR, old_db);
+}
+
 static void vmx_db_test_guest(void)
 {
 	/*
@@ -9623,6 +9742,7 @@ struct vmx_test vmx_tests[] = {
 	TEST(vmx_pending_event_test),
 	TEST(vmx_pending_event_hlt_test),
 	TEST(vmx_store_tsc_test),
+	TEST(vmx_preemption_timer_zero_test),
 	/* EPT access tests. */
 	TEST(ept_access_test_not_present),
 	TEST(ept_access_test_read_only),

From patchwork Tue Apr 14 00:10:26 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 11486481
Return-Path: <SRS0=cOhZ=56=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 1B02D92C
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 14 Apr 2020 00:10:36 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id F308A2072D
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 14 Apr 2020 00:10:35 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=google.com header.i=@google.com
 header.b="RXm1S12r"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S2390370AbgDNAKe (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 13 Apr 2020 20:10:34 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:44792 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-FAIL-OK-FAIL)
        by vger.kernel.org with ESMTP id S1727878AbgDNAKd (ORCPT
        <rfc822;kvm@vger.kernel.org>); Mon, 13 Apr 2020 20:10:33 -0400
Received: from mail-pj1-x104a.google.com (mail-pj1-x104a.google.com
 [IPv6:2607:f8b0:4864:20::104a])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 13F08C0A3BDC
        for <kvm@vger.kernel.org>; Mon, 13 Apr 2020 17:10:33 -0700 (PDT)
Received: by mail-pj1-x104a.google.com with SMTP id x6so3479467pjg.5
        for <kvm@vger.kernel.org>; Mon, 13 Apr 2020 17:10:33 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=bVvnLr/13Rl1XJjZsivRn7ggZ8MLXPQ/yl/t3ThpoaI=;
        b=RXm1S12ruTwKRtz7mCQPgTkIPLoTzGOaiUL3LDnLkW2X0qOKL4+OzFij0qYp0OKlmt
         KCQhkf8bLnh4cHQcbLg+OPkwhlQwuycW8FNF5Arultd9v7Ea0zq1/ZPY+UEDHnaNXeuy
         LQJlA/3MrCSi9ElipRGMHJcA9yhWLeJdGlz2NSUaDakxRSEK+uvym9jbaLtviztJfi8O
         hfiM/nqIa0L4H7qvj+Pb0o9RUeXTcPtCVQKiEtvce6RckNLDWK5LejszauzwlCSIXcYT
         sf7/1pp7QV1Nvuf3jnGO7v+wxMgqMtZm7aL6va5FV2Vi/sRRYEoEzjm4iFKKRDemgDDq
         kxfQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=bVvnLr/13Rl1XJjZsivRn7ggZ8MLXPQ/yl/t3ThpoaI=;
        b=HqdUkg50utgZeyxeWZ5cMgb8Jj4brceJWXJnfuC1h/pVNiR4GQiR7k8gxDhjBRNH0b
         AFYnIjospLSpCnVYtIpwP8fR1jqkuR5Oa88U0ewL/P2UEqquIADHiMx250sYq48VoHTt
         8VjfU9MIK3x8XBnSCNcX1DfB3eb3+STp9kA9k/Oe5sf7WtEWRNZKfJwThguuPqWCNOrt
         aaitxuYpcaIoMxz4njnkmhBOr7QGLvVKXFQQ/CQ3hU6xl8MGrojHC78Nq0OJfu1Du2LA
         sVvHA99yo021deJBtXueT2osvBdzjRD/MJywy6kx8I/G8rN9gJ0UuF0qwGu/5Iw8trvL
         vX1A==
X-Gm-Message-State: AGi0PubttyFZ5qJfV5mc3+0xASn+u7cT4Y9C4v+Q0nrVdADvrWe9I6RL
        QX9Ml2CAmPO/rdUrhO54Vud7jKvjQxaVygoExOdY8DxQaQrLqNSl6886T2YWSiI5wwfPH+hcLFQ
        eZeWAavZIjzdgfXTCVADrrfE/+MpiWb6NpqTqOEE45aPNnqJWaNjwC4gHTB++FGU=
X-Google-Smtp-Source: 
 APiQypJn+ICm4+7HDZKJB5RGhvwH2+IEzBAMHuBnGyf8jJqlEXJUKEtJLbOfaxB7j4bFoYsoEAxTYTzUJ5PELQ==
X-Received: by 2002:a17:90a:c392:: with SMTP id
 h18mr25032688pjt.89.1586823032394;
 Mon, 13 Apr 2020 17:10:32 -0700 (PDT)
Date: Mon, 13 Apr 2020 17:10:26 -0700
In-Reply-To: <20200414001026.50051-1-jmattson@google.com>
Message-Id: <20200414001026.50051-2-jmattson@google.com>
Mime-Version: 1.0
References: <20200414001026.50051-1-jmattson@google.com>
X-Mailer: git-send-email 2.26.0.110.g2183baf09c-goog
Subject: [kvm-unit-tests PATCH 2/2] x86: VMX: Add another corner-case
 VMX-preemption timer test
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>,
        Oliver Upton <oupton@google.com>,
        Peter Shier <pshier@google.com>
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Ensure that the delivery of a "VMX-preemption timer expired" VM-exit
doesn't disrupt single-stepping in the guest. Note that passing this
test doesn't ensure correctness.

Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Oliver Upton <oupton@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
---
 x86/vmx_tests.c | 104 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 104 insertions(+)

diff --git a/x86/vmx_tests.c b/x86/vmx_tests.c
index fccb27f..86b8880 100644
--- a/x86/vmx_tests.c
+++ b/x86/vmx_tests.c
@@ -8438,6 +8438,109 @@ static void vmx_preemption_timer_zero_test(void)
 	handle_exception(DB_VECTOR, old_db);
 }
 
+static u64 vmx_preemption_timer_tf_test_prev_rip;
+
+static void vmx_preemption_timer_tf_test_db_handler(struct ex_regs *regs)
+{
+	extern char vmx_preemption_timer_tf_test_endloop;
+
+	if (vmx_get_test_stage() == 2) {
+		/*
+		 * Stage 2 means that we're done, one way or another.
+		 * Arrange for the iret to drop us out of the wbinvd
+		 * loop and stop single-stepping.
+		 */
+		regs->rip = (u64)&vmx_preemption_timer_tf_test_endloop;
+		regs->rflags &= ~X86_EFLAGS_TF;
+	} else if (regs->rip == vmx_preemption_timer_tf_test_prev_rip) {
+		/*
+		 * The RIP should alternate between the wbinvd and the
+		 * jmp instruction in the code below. If we ever see
+		 * the same instruction twice in a row, that means a
+		 * single-step trap has been dropped. Let the
+		 * hypervisor know about the failure by executing a
+		 * VMCALL.
+		 */
+		vmcall();
+	}
+	vmx_preemption_timer_tf_test_prev_rip = regs->rip;
+}
+
+static void vmx_preemption_timer_tf_test_guest(void)
+{
+	/*
+	 * The hypervisor doesn't intercept WBINVD, so the loop below
+	 * shouldn't be a problem--it's just two instructions
+	 * executing in VMX non-root mode. However, when the
+	 * hypervisor is running in a virtual environment, the parent
+	 * hypervisor might intercept WBINVD and emulate it. If the
+	 * parent hypervisor is broken, the single-step trap after the
+	 * WBINVD might be lost.
+	 */
+	asm volatile("vmcall\n\t"
+		     "0: wbinvd\n\t"
+		     "1: jmp 0b\n\t"
+		     "vmx_preemption_timer_tf_test_endloop:");
+}
+
+/*
+ * Ensure that the delivery of a "VMX-preemption timer expired"
+ * VM-exit doesn't disrupt single-stepping in the guest. Note that
+ * passing this test doesn't ensure correctness, because the test will
+ * only fail if the VMX-preemtion timer fires at the right time (or
+ * the wrong time, as it were).
+ */
+static void vmx_preemption_timer_tf_test(void)
+{
+	handler old_db;
+	u32 reason;
+	int i;
+
+	if (!(ctrl_pin_rev.clr & PIN_PREEMPT)) {
+		report_skip("'Activate VMX-preemption timer' not supported");
+		return;
+	}
+
+	old_db = handle_exception(DB_VECTOR,
+				  vmx_preemption_timer_tf_test_db_handler);
+
+	test_set_guest(vmx_preemption_timer_tf_test_guest);
+
+	enter_guest();
+	skip_exit_vmcall();
+
+	vmx_set_test_stage(1);
+	vmcs_set_bits(PIN_CONTROLS, PIN_PREEMPT);
+	vmcs_write(PREEMPT_TIMER_VALUE, 50000);
+	vmcs_write(GUEST_RFLAGS, X86_EFLAGS_FIXED | X86_EFLAGS_TF);
+
+	/*
+	 * The only exit we should see is "VMX-preemption timer
+	 * expired."  If we get a VMCALL exit, that means the #DB
+	 * handler has detected a missing single-step trap. It doesn't
+	 * matter where the guest RIP is when the VMX-preemption timer
+	 * expires (whether it's in the WBINVD loop or in the #DB
+	 * handler)--a single-step trap should never be discarded.
+	 */
+	for (i = 0; i < 10000; i++) {
+		enter_guest();
+		reason = (u32)vmcs_read(EXI_REASON);
+		if (reason == VMX_PREEMPT)
+			continue;
+		TEST_ASSERT(reason == VMX_VMCALL);
+		skip_exit_insn();
+		break;
+	}
+
+	report(reason == VMX_PREEMPT, "No single-step traps skipped");
+
+	vmx_set_test_stage(2);
+	vmcs_clear_bits(PIN_CONTROLS, PIN_PREEMPT);
+	enter_guest();
+
+	handle_exception(DB_VECTOR, old_db);
+}
+
 static void vmx_db_test_guest(void)
 {
 	/*
@@ -9743,6 +9846,7 @@ struct vmx_test vmx_tests[] = {
 	TEST(vmx_pending_event_hlt_test),
 	TEST(vmx_store_tsc_test),
 	TEST(vmx_preemption_timer_zero_test),
+	TEST(vmx_preemption_timer_tf_test),
 	/* EPT access tests. */
 	TEST(ept_access_test_not_present),
 	TEST(ept_access_test_read_only),
